examples/job-shop-scheduling/jobShopScheduler.py

```diff
@@ -0,0 +1,268 @@
+                for other_task in same_machine_tasks:
+                    if task.job == other_task.job and task.position == other_task.position:
+                        continue
+
+                    for t in range(self.max_time):
+                        current_label = get_label(task, t)
+
+                        for tt in range(t, min(t + task.duration, self.max_time)):
+                            self.csp.add_constraint(valid_values, {current_label,
+                                                                   get_label(other_task, tt)})
+
+    def _remove_absurd_times(self):
+        """Sets impossible task times in self.csp to 0.
+        """
+        # TODO: deal with overlaps in time
+        for task in self.tasks:
+            # Times that are too early for task
+            for t in range(task.position):
+                label = get_label(task, t)
+                self.csp.fix_variable(label, 0)
+
+            # Times that are too late for task to complete
+            for t in range(task.duration - 1):  # -1 to ignore duration==1
+                label = get_label(task, (self.max_time - 1) - t)  # -1 for zero-indexed time
+                self.csp.fix_variable(label, 0)
+
+    def get_bqm(self):
+        """Returns a BQM to the Job Shop Scheduling problem.
+
+        Example on usage:
+            # 'jobs' dict describes the jobs we're interested in scheduling. Namely,
+            # the dict key is the name of the job and the dict value is the ordered
+            # list of tasks that the job must do.
+            #
+            # {"job_name": [(machine_name, time_duration_on_machine), ..],
+            #  "another_job_name": [(some_machine, time_duration_on_machine), ..]}
+
+            jobs = {"job_a": [("mach_1", 2), ("mach_2", 2), ("mach_3", 2)],
+                    "job_b": [("mach_3", 3), ("mach_2", 1), ("mach_1", 1)],
+                    "job_c": [("mach_2", 2), ("mach_1", 3), ("mach_2", 1)]}
+            max_time = 6	# Put an upperbound on how long the schedule can be
+
+            jss = JobShopScheduler(jobs, max_time)
+            jss.get_bqm()   # Run job shop scheduling constraints
+        """
+        # Apply constraints to self.csp
+        self._add_one_start_constraint()
+        self._add_precedence_constraint()
+        self._add_share_machine_constraint()
+        self._remove_absurd_times()
+
+        # Get BQM
+        bqm = dwavebinarycsp.stitch(self.csp)
+
+        # Edit BQM
+        base = len(self.last_task_indices) + 1     # Base for exponent
+        for i in self.last_task_indices:
+            task = self.tasks[i]
+
+            for t in range(self.max_time):
+                end_time = t + task.duration
+
+                # Check task's end time; do not add in absurd times
+                if end_time > self.max_time:
+                    continue
+
+                # Add bias to variable
+                bias = 2 * base**(end_time - self.max_time)
```

Comment:
Looks like this will escalate pretty quickly even for small problems. Even if finite resolution wasn't a problem, auto-scaling would practically annul other constraints, no?

Reply:
Yes, the bias will probably need to be re-scaled in order to prevent it from overshadowing the other constraints.

It escalates quickly because I wanted a penalty scheme that would mathematically guarantee that the smallest penalty would always correspond to the smallest schedule time. Having a base of `N+1` and a geometric penalty allows me to achieve that.

I should probably have a comment on how this bias came about. Essentially, worst case optimal solution penalty < any non-optimal solution penalty.

Reply:
Out of curiosity, I wonder what's the largest `base` and largest `max_time` we can effectively handle on our system with this exponential penalty scaling. When I say effectively handle, I mean to get a valid solution in let's say 50% of the time, and to get an optimal solution at least in 1% of the samples.

Reply:
Hi @randomir, you asked a really great question. I don't know about effective handling with this penalty scaling, but I do worry about fitting the problem on the QPU. I have some upper bounds below. The bounds below have not accounted for chaining.

```
N_tasks = total number of tasks in the schedule
T = max time allowed in schedule

# nodes in graph = T * N_tasks
# edges in graph: O(T^2 * N_tasks) + O(T * N_tasks^2)

Specific edge break down:
# edges from one start constraint:  O(T^2 * N_tasks)
# edges from precedence constraint: O(T * N_tasks)
# edges from machine constraint: O(T * N_tasks^2) 
```

Reply:
I've put an issue to look into scaling/changing this shortest schedule penalty. See #9

Reply:
Thanks Mel for those detailed bounds. I’ll stay tuned to #9 for some benchmarking results. :)